bson/bsoncodec/registry.go

```diff
@@ -424,13 +463,14 @@ func (r *Registry) lookupTypeDecoder(t reflect.Type) (ValueDecoder, bool) {
 	return dec, found
 }
 
-func (r *Registry) lookupInterfaceDecoder(t reflect.Type) (ValueDecoder, bool) {
+func (r *Registry) lookupInterfaceDecoder(t reflect.Type, allowAddr bool) (ValueDecoder, bool) {
 	for _, idec := range r.interfaceDecoders {
-		if !t.Implements(idec.i) && !reflect.PtrTo(t).Implements(idec.i) {
-			continue
+		if t.Implements(idec.i) {
+			return idec.vd, true
+		}
+		if allowAddr && t.Kind() != reflect.Ptr && reflect.PtrTo(t).Implements(idec.i) {
```

Comment:
Why isn't a recursive call to `lookupInterfaceDecoder(t, false)` needed here as it is in `lookupInterfaceEncoder`?

Reply:
Same thing here about leaving a comment. Also, I'm not sure if the same logic is needed for the decode code path. The top-level `Unmarshal` functions only allow a pointer or map to be specified. Does `encoding/json` have a conditional decoder type?

Reply:
encoding/json has a TextUnmarshaler too